contracts/staking/ValidatorManager.sol

```diff
@@ -93,6 +90,105 @@ abstract contract ValidatorManager is
         $._pChainBlockchainID = settings.pChainBlockchainID;
         $._subnetID = settings.subnetID;
         $._maximumHourlyChurn = settings.maximumHourlyChurn;
+        $._initializedValidatorSet = false;
+    }
+
+    modifier initializedValidatorSet() {
+        require(
+            _getValidatorManagerStorage()._initializedValidatorSet,
+            "ValidatorManager: validator set not initialized"
+        );
+        _;
+    }
+
+    /**
+     * @notice See {IValidatorManager-initializeValidatorSet}.
+     */
+    function initializeValidatorSet(
+        SubnetConversionData calldata subnetConversionData,
+        uint32 messageIndex
+    ) external {
+        ValidatorManagerStorage storage $ = _getValidatorManagerStorage();
+        require(!$._initializedValidatorSet, "ValidatorManager: already initialized validator set");
+        require(
+            subnetConversionData.blockchainID == WARP_MESSENGER.getBlockchainID(),
+            "ValidatorManager: invalid blockchain ID"
+        );
+        require(
+            address(subnetConversionData.validatorManagerAddress) == address(this),
+            "ValidatorManager: invalid validator manager address"
+        );
+        require(
+            subnetConversionData.initialValidators.length > 0,
+            "ValidatorManager: no initial validators"
+        );
+        uint256 numInitialValidators = subnetConversionData.initialValidators.length;
+        require(
+            numInitialValidators < type(uint32).max, "ValidatorManager: too many initial validators"
+        );
+
+        // Verify that the sha256 hash of the Subnet conversion data matches with the Warp message's subnetConversionID.
+        bytes memory encodedConversion = abi.encodePacked(
+            subnetConversionData.convertSubnetTxID,
+            subnetConversionData.blockchainID,
+            uint32(20),
+            subnetConversionData.validatorManagerAddress,
+            uint32(numInitialValidators)
+        );
+        for (uint256 i; i < numInitialValidators; i++) {
+            InitialValidator memory initialValidator = subnetConversionData.initialValidators[i];
+            bytes32 nodeID = initialValidator.nodeID;
+            require(nodeID != bytes32(0), "ValidatorManager: invalid node ID");
+            require(
+                $._activeValidators[nodeID] == bytes32(0),
+                "ValidatorManager: node ID already active"
+            );
+            require(
+                initialValidator.blsPublicKey.length == 48,
+                "ValidatorManager: invalid blsPublicKey length"
+            );
+
+            // Continue to encode the initial validators.
+            encodedConversion = abi.encodePacked(
```

Comment:
I think this results in a copy of the full `encodedConversion` bytes up until this point. It'd probably be a bit cleaner and more efficient to pre-allocate an array of the necessary size upfront once, and then assign into it as we go along.